sycl/source/detail/kernel_info.hpp

```diff
@@ -71,8 +71,22 @@ typename std::enable_if<!IsSubGroupInfo<Param>::value>::type
 get_kernel_device_specific_info_helper(RT::PiKernel Kernel, RT::PiDevice Device,
                                        const plugin &Plugin, void *Result,
                                        size_t Size) {
-  Plugin.call<PiApiKind::piKernelGetGroupInfo>(
-      Kernel, Device, PiInfoCode<Param>::value, Size, Result, nullptr);
+  try {
+    Plugin.call<PiApiKind::piKernelGetGroupInfo>(
+        Kernel, Device, PiInfoCode<Param>::value, Size, Result, nullptr);
+  } catch (sycl::exception &e) {
+    // clGetKernelWorkGroupInfo returns CL_INVALID_VALUE if param_name is
+    // CL_KERNEL_GLOBAL_WORK_SIZE and device is not a custom device and kernel
+    // is not a built-in kernel. According to SYCL2020 an exception with the
+    // errc::invalid error code should be thrown.
+    std::string msg = e.what();
+    if (msg.find("PI_ERROR_INVALID_VALUE") != std::string::npos)
+      throw sycl::exception(
+          sycl::make_error_code(errc::invalid),
+          "info::kernel_device_specific::global_work_size descriptor may only "
+          "be used if the device type is device_type::custom or if the kernel "
+          "is a built-in kernel.");
+  }
```

Comment:
Instead of letting it throw first, you could use `call_nocheck` and check the returned `PI_ERROR_*` first, then if it isn't one of the ones we want to handle explicitly we call `checkPiResult`. I.e. something like:
```suggestion
  RT::PiResult Err = Plugin.call_nocheck<PiApiKind::piKernelGetGroupInfo>(
      Kernel, Device, PiInfoCode<Param>::value, Size, Result, nullptr);
  // clGetKernelWorkGroupInfo returns CL_INVALID_VALUE if param_name is
  // CL_KERNEL_GLOBAL_WORK_SIZE and device is not a custom device and kernel
  // is not a built-in kernel. According to SYCL2020 an exception with the
  // errc::invalid error code should be thrown.
  if (Err == PI_ERROR_INVALID_VALUE)
    throw sycl::exception(
      sycl::make_error_code(errc::invalid),
      "info::kernel_device_specific::global_work_size descriptor may only "
      "be used if the device type is device_type::custom or if the kernel "
      "is a built-in kernel.");
  Plugin.checkPiResult(Err);
```

Reply:
Do we check somewhere if `param_name` is actually `CL_KERNEL_GLOBAL_WORK_SIZE`? Because if not, then it looks weird to me. `CL_INVALID_VALUE` may be returned for other queries as well.

Please also note that it may be better to outline error code to exception conversion into a separate dedicated file, see #860 as an example

Reply:
`clGetKernelWorkGroupInfo` can return `CL_INVALID_VALUE` in two cases:

> CL_INVALID_VALUE if param_name is not valid, or if size in bytes specified by param_value_size is < size of return type as described in the table above and param_value is not NULL.
> CL_INVALID_VALUE if param_name is CL_KERNEL_GLOBAL_WORK_SIZE and device is not a custom device and kernel is not a built-in kernel.

See: https://registry.khronos.org/OpenCL/sdk/2.1/docs/man/xhtml/clGetKernelWorkGroupInfo.html

I thought that we can't get the first case as our sycl implementation just won't compile a call with the wrong parameters, or no? So I decided we don't have to check it

Reply:
> I thought that we can't get the first case as our sycl implementation just won't compile a call with the wrong parameters, or no? So I decided we don't have to check it

Strictly speaking, we are handling result of `piKernelGetGroupInfo` which may not be equivalent to `clGetKernelWorkGroupInfo` depending on the backend/plugin used. However, the current implementation is probably fine, because that is just an error handling for some corner case